### ISSUE TITLE
Embedding branch 问题？

### ISSUE THREAD

**Issue by SentimentSongs** [MaybeShewill-CV/lanenet-lane-detection#67]
用的数据是 tusimple lane dataset
训练出来的embedding.png 显示出来的照片一直都是 雪花点，不是正常的图片
论文上 Embedding branch。您这块代码是不是有问题，或者没有这块代码？

进行聚合的时候 出现nan的问题？，计算loss的时候 被除数肯定是0的情况下？
我看了所有的图片，二值图都是正常的，实例图也是没有问题。。

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs embedding branch原始论文中的embedding是4维特征向量，我这里的实现了embedding到了3维特征向量，embedding结果可以参照readme的embedding示例图像。至于loss等于nan的情况可能是超参数不对的问题。可以参考前面的问题#33

**Comment by SentimentSongs** [comment]
test_的时候调了lanenet_cluster.py  得到一个 mask_image。。

这个mask_image 是不是train需要二值图？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 使用lancet_cluster.py是因为特征图需要聚类来生成不同的车道线。mask_image是不是train需要二值图是什么意思？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs instance label是参与实例分隔的损失计算。Instance label直接参与lanenet_discriminative_loss.discriminative_loss你觉得会出现的问题是什么呢？

**Comment by SentimentSongs** [comment]
lanenet_merge_model里边的compute_loss方法 只有计算二值分割损失
 实例分割的损失计算在哪里？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs lanenet_merge_model.py的114-118行就是计算实例分割损失的 你可以再仔细看一下：）

**Comment by SentimentSongs** [comment]
原论文的 cluster 这一步， 在train的时候 在怎么聚类的？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs mean shift过程并不参与网络训练 这个你可能搞错作者的意思了:)

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 整个网络的loss由两部分构成， 一部分是binary segmentation loss， 另一部分是instance segmentation loss。这两部分loss和你上面说的mean shift cluster都没有关系。mean shift cluster是预测的时候才会用到的

**Comment by SentimentSongs** [comment]
![binary_seg_img](https://user-images.githubusercontent.com/43947279/46651936-2d668980-cbd4-11e8-8df7-8615025d4ce5.png)
train 训练中binary_seg_img 为什么会出现这的图片

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 初始迭代的时候网络近似于随机预测所以会出现这样的情况 可以等训练一段时间再观察

**Comment by SentimentSongs** [comment]
训练到1万 以上 还会出现这样的图片

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 那可能是超参数设置问题。可以贴一下你完整的loss曲线截图

**Comment by SentimentSongs** [comment]
tensorboard  一直打不开,
![1539069145094](https://user-images.githubusercontent.com/43947279/46652490-c3e77a80-cbd5-11e8-9c2c-c7aabb139d4e.jpg)

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 贴一下完整的控制台截图

**Comment by SentimentSongs** [comment]
![wx20181009-154619](https://user-images.githubusercontent.com/43947279/46654103-7d484f00-cbda-11e8-8d85-74b359567985.png)

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 默认的tensorboard文件存放在tboard文件夹下 你可能需要检查一下tensorboard相关文件存放路径或者google一下tensorboard的用法：）

**Comment by SentimentSongs** [comment]
刷新下

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 按照这个曲线不应该出现类似随机噪声点的binary segmentation结果 你可以加载网络训练的中间结果测试看下binary segmentation的结果

**Comment by SentimentSongs** [comment]
是什么原因会造成随机噪声点

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 网络参数没有收敛

**Comment by SentimentSongs** [comment]
Epoch 到1万以上
cost is
binary cost is
instance cost is
 三个就出现了nan。。

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 这个需要调整超参数 可以参考前面的问题 #33

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
参数除了 __C.TRAIN.EPOCHS = 200010
__C.TRAIN.TEST_DISPLAY_STEP = 1000
__C.TRAIN.LEARNING_RATE = 0.0005
__C.TRAIN.LR_DECAY_STEPS = 210000

 还调什么参数和训练相关的

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 还有在计算instance segmentation的时候需要确定的阈值 你可以查看lanenet_merge_model.py文件的114-118行

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 没有明白意思？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 那几行代码函数中传入的参数就是instance segmentation损失计算部分需要的超参数，这部分你可以对照原论文查看一下

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV  多少算超参，或者跟global_config文件里边那个参数做比较？？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs ......超参数不是说超过了某个参数的意思==！ 它的意思你可以简单理解为需要人工按照经验指定的参数：）

**Comment by SentimentSongs** [comment]
使用VGG16 作为主干的话

1.您使用了整个vgg16的模型，还是用了几层，又加了别的模型？
2.lanenet_binary_segmentation 和 lanenet_instance_segmentation 使用了三层 pool
 binary  和 instance 怎么聚合的？ 这点我看论文不是太明白

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 1.你查看下代码很清楚的写了用了几层。2.你说的3层pool binary是什么意思我没明白

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 1.代码里面注释很清楚， 用了vgg前五个stage 2.这个意思不是你说的三层pool binary，意思是decode了这三层来恢复原图像的信息

**Comment by ktr-hubrt** [comment]
作者你好，我看论文中说binary分支和embedding分支是两个独立的branch，但是我看你的实现好像，并不是独立的两只，而是把binary的预测结果作为embedding的输入然后接了一个卷积一个relu？

**Comment by MaybeShewill-CV** [comment]
@ktr-hubrt 可以查看下lanenet_merge_model.py第111行, embedding是从encoder的结果开始的，也就是原论文中的独立分支.你也可以看下readme中的网络结构图，有助于你理解。

**Comment by SentimentSongs** [comment]
我用的 tusimple lane dataset 数据集
但是代码怎么跑出来  在culane的文件夹  和culane_lanenet_vgg的ckpt

 怎么不是tusimple文件夹和 tusimpl_lanenet_vgg的ckpt

名字不同，运行的代码有什么影响嘛？

**Comment by ktr-hubrt** [comment]
> ![wx20181018-091605](https://user-images.githubusercontent.com/43947279/47125524-afcb0980-d2b6-11e8-8648-3743bd6281ab.png)
> 原图都是训练过的。
> 我是我训练108000生成的ckpt，来做验证的结果
> 
> mask_image 和binary_image怎么是全黑的，没有车道线

你把binary_image像素*100输出看看呢

**Comment by ktr-hubrt** [comment]
> @ktr-hubrt 怎么操作？

binary_image = binary_image*100
print（binary_image）

**Comment by SentimentSongs** [comment]
@ktr-hubrt 
代码 test_lanenet.py  135行的 
 plt.imshow(binary_seg_image[0] * 255, cmap='gray')
我改成了 
 plt.imshow(binary_seg_image[0] * 100, cmap='gray')

这样行不

**Comment by SentimentSongs** [comment]
I1018 11:45:27.221402 47539 test_lanenet.py:102] 单张图像车道线预测耗时: 3.49740s
E1018 11:45:27.239267 47539 lanenet_cluster.py:58] Expected 2D array, got 1D array instead:
array=[].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

**Comment by ktr-hubrt** [comment]
> I1018 11:45:27.221402 47539 test_lanenet.py:102] 单张图像车道线预测耗时: 3.49740s
> E1018 11:45:27.239267 47539 lanenet_cluster.py:58] Expected 2D array, got 1D array instead:
> array=[].
> Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

不是像素值的问题。。。。

**Comment by ktr-hubrt** [comment]
> I1018 11:45:27.221402 47539 test_lanenet.py:102] 单张图像车道线预测耗时: 3.49740s
> E1018 11:45:27.239267 47539 lanenet_cluster.py:58] Expected 2D array, got 1D array instead:
> array=[].
> Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

那不是像素值（显示）的问题了，你看看训练时输入网络的binary label对不对吧，你这没学到啊

**Comment by SentimentSongs** [comment]
@ktr-hubrt 
lanenet_instance_segmentation.py
99行
 pix_embedding, label, 3, image_shape, 0.5, 1.5, 1.0, 1.0, 0.001)
我改成了
 pix_embedding, label, 3, image_shape, 0.5, 3, 1.0, 1.0, 0.001)
训练的有影响嘛？

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
![0000](https://user-images.githubusercontent.com/43947279/47199795-f51a3480-d3a5-11e8-8aa2-40e57c34dca9.PNG)

accuracy  为什么会出现0.000的情况

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 网络收敛需要时间

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
![wx20181018-091605](https://user-images.githubusercontent.com/43947279/47125524-afcb0980-d2b6-11e8-8648-3743bd6281ab.png)
原图都是训练过的。
我是我训练108000生成的ckpt，来做验证的结果

mask_image 和binary_image怎么是全黑的，没有车道线

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 调整学习率等超参数重新训练

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 学习率 我训练的0.0001
我改成0.0005 看看效果

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 也可以尝试别的optimizer

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 这个就是训练失败啊==！

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
我的意思是 训练时输入网络的binary label不对
还是 调整学习率等超参数的问题?

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 你说的两个点都有可能造成训练失败 标签是否正确需要你自己检查别人帮不了你

**Comment by SentimentSongs** [comment]
训练了20万遍

test验证cpkt

mask_image 和binary_image 还是全部黑色

但是在训练的时候 binary_label.png
![binary_label](https://user-images.githubusercontent.com/43947279/47275789-f6857000-d5e4-11e8-93f0-946b63e85683.png)

binary_seg_img.png
![binary_seg_img](https://user-images.githubusercontent.com/43947279/47275793-056c2280-d5e5-11e8-9a2d-548955bcdf43.png)
embedding.png
![embedding](https://user-images.githubusercontent.com/43947279/47275799-11f07b00-d5e5-11e8-9e13-ecfd4820fe38.png)

instance_label.png

![instance_label](https://user-images.githubusercontent.com/43947279/47275801-216fc400-d5e5-11e8-8a7d-ea926e81bcaa.png)
image.png

![image](https://user-images.githubusercontent.com/43947279/47275809-30ef0d00-d5e5-11e8-90dc-0b727387aaa6.png)


是否正确

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs Binary segmentation result is correct but the embedding result is not correct. You may check whether you have changed the embedding feature dimensions or the embedding result has not been correctly scaled.

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs Check if you wrote the wrong file path in train.txt file. You can use the absolute
 path instead.

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
txt文件用的就是绝对路径
train_dataset.next_batch 读的self._gt_label_binary_list
self._gt_label_instance_lis
self._gt_img_list

三个list都都路径没问题的

return gt_imgs 图片是没问题的

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs Last time you said the gt_labels_binary list is empty so I want to know whether it is empty or not==!

**Comment by SentimentSongs** [comment]
train_lanenet.py 数据集culane
train_lanenet_instance_segmentation.py  数据集tusimple
train_encoder_decoder.py 数据集kitti

都是train 但是对应数据集不同 ？？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs The training script is just the same no matter what dataset you use. The train_lanenet_instance_segmentation.py and train_encoder_decoder.py are not used in this model and I have deleted them recently you may pull the latest code.

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
我已经训练出来cpkt 
我拿test_lanenet.py来验证结果 

1.我经过查看代码  lanenet_postprocess.py 这里边的方法
def postprocess(self, image, minarea_threshold=200):
stats[4]这个数组的值都是小于200的 所以 mask_image是空的
我改成1了mask_image有线了
2.lanenet_cluster.py类中
    def get_lane_mask(self, binary_seg_ret, instance_seg_ret):
coord = np.array([coord])
以下图片我是用cv2.polylines方法用的coord数据的坐标
![wechatimg88](https://user-images.githubusercontent.com/43947279/47490620-b25fcd00-d87b-11e8-9875-191e285a9dbd.png)
![wechatimg87](https://user-images.githubusercontent.com/43947279/47490623-b4299080-d87b-11e8-91e9-26bc6296a46b.png)
![wechatimg86](https://user-images.githubusercontent.com/43947279/47490626-b55abd80-d87b-11e8-9fe9-b2aa1ce893a2.png)

出现了不是车道线的线，， 这个是什么原因造成？？？

不知道您能不能明白我的意思？？

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 这个中间的连线是因为错误的车道线聚类结果造成的。正常情况下，binary mask会提供车道线的掩码点，这些点会根据embedding的feature来进行聚类，正确情况下，同一根车道线上的点会被聚类成一类 不同车道线上的点会被聚类成不同类 然后同类的点链接起来就会生成车道线。你上图所示的错误情况就是因为车道线点聚类错了，所以不同车道线上的点被错误的链接起来了。

**Comment by SentimentSongs** [comment]
![1541387447890](https://user-images.githubusercontent.com/43947279/47976031-81895e80-e0eb-11e8-94ea-69ad03db2568.jpg)

有跟线有延长的部分， 是什么原因造成？
我看了输出的数组 是4个

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 聚类点有错分 这个你需要自己增加策略删除异常点

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 
有什么思路嘛，

**Comment by SentimentSongs** [comment]
@MaybeShewill-CV 我现在主要是不明白，您这边是怎么聚类点的，所以我现在也不明白该怎么去做聚类分析

**Comment by MaybeShewill-CV** [comment]
@SentimentSongs 根据embedding生成的feature用mean shift方法聚类 具体方法可以参照论文

**Comment by zacario-li** [comment]
> @MaybeShewill-CV
> ![0000](https://user-images.githubusercontent.com/43947279/47199795-f51a3480-d3a5-11e8-8aa2-40e57c34dca9.PNG)
> 
> accuracy 为什么会出现0.000的情况

accuracy是0的问题，你解决了没？

**Comment by MaybeShewill-CV** [comment]
@zacario-li I do not have such problem:)
